dynamodb/src/main/java/org/apache/metamodel/dynamodb/DynamoDbDataContext.java

```diff
@@ -0,0 +1,320 @@
+     * creating new tables. Defaults to 5.
+     */
+    public static final String SYSTEM_PROPERTY_THROUGHPUT_WRITE_CAPACITY = "metamodel.dynamodb.throughput.capacity.write";
+
+    private static final Logger logger = LoggerFactory.getLogger(DynamoDbDataContext.class);
+
+    /**
+     * The artificial schema name used by this DataContext.
+     */
+    public static final String SCHEMA_NAME = "public";
+
+    private final AmazonDynamoDB _dynamoDb;
+    private final boolean _shutdownOnClose;
+    private final SimpleTableDef[] _tableDefs;
+
+    public DynamoDbDataContext() {
+        this(AmazonDynamoDBClientBuilder.defaultClient(), null, true);
+    }
+
+    public DynamoDbDataContext(SimpleTableDef[] tableDefs) {
+        this(AmazonDynamoDBClientBuilder.defaultClient(), tableDefs, true);
+    }
+
+    public DynamoDbDataContext(AmazonDynamoDB client) {
+        this(client, null, false);
+    }
+
+    public DynamoDbDataContext(AmazonDynamoDB client, SimpleTableDef[] tableDefs) {
+        this(client, tableDefs, false);
+    }
+
+    private DynamoDbDataContext(AmazonDynamoDB client, SimpleTableDef[] tableDefs, boolean shutdownOnClose) {
+        _dynamoDb = client;
+        _tableDefs = (tableDefs == null ? new SimpleTableDef[0] : tableDefs);
+        _shutdownOnClose = shutdownOnClose;
+    }
+
+    public AmazonDynamoDB getDynamoDb() {
+        return _dynamoDb;
+    }
+
+    @Override
+    public void close() {
+        if (_shutdownOnClose) {
+            _dynamoDb.shutdown();
+        }
+    }
+
+    @Override
+    protected Schema getMainSchema() throws MetaModelException {
+        final Map<String, SimpleTableDef> tableDefs = new HashMap<>();
+        for (SimpleTableDef tableDef : _tableDefs) {
+            tableDefs.put(tableDef.getName(), tableDef);
+        }
+
+        final MutableSchema schema = new MutableSchema(getMainSchemaName());
+        final ListTablesResult tables = _dynamoDb.listTables();
+        final List<String> tableNames = tables.getTableNames();
+        for (String tableName : tableNames) {
+            final MutableTable table = new MutableTable(tableName, schema);
+            schema.addTable(table);
+
+            final DescribeTableResult descripeTableResult = _dynamoDb.describeTable(tableName);
+            final TableDescription tableDescription = descripeTableResult.getTable();
+
+            // add primary keys
+            addColumnFromKeySchema("Primary index", tableDescription.getKeySchema(), table, true);
+
+            // add attributes from global and local indices
+            final List<GlobalSecondaryIndexDescription> globalSecondaryIndexes = tableDescription
+                    .getGlobalSecondaryIndexes();
+            if (globalSecondaryIndexes != null) {
+                for (GlobalSecondaryIndexDescription globalSecondaryIndex : globalSecondaryIndexes) {
+                    addColumnFromKeySchema(globalSecondaryIndex.getIndexName(), globalSecondaryIndex.getKeySchema(),
```

Comment:
Unfortunately the GlobalSecondaryIndexDescription and LocalSecondaryIndexDescription classes don't share a common super-class, so I'm not sure if it can be made more general.